app/src/main/java/org/xbmc/kore/ui/sections/remote/ShareHandlingFragment.java

```diff
@@ -0,0 +1,370 @@
+import android.support.annotation.Nullable;
+import android.support.annotation.StringRes;
+import android.support.v4.app.Fragment;
+import android.support.v4.app.FragmentManager;
+import android.widget.Toast;
+
+import org.xbmc.kore.R;
+import org.xbmc.kore.jsonrpc.ApiCallback;
+import org.xbmc.kore.jsonrpc.ApiException;
+import org.xbmc.kore.jsonrpc.HostConnection;
+import org.xbmc.kore.jsonrpc.method.Player;
+import org.xbmc.kore.jsonrpc.method.Playlist;
+import org.xbmc.kore.jsonrpc.type.PlayerType.GetActivePlayersReturnType;
+import org.xbmc.kore.jsonrpc.type.PlaylistType;
+
+import java.util.ArrayList;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Headless fragment that sends shared URLs to Kodi plugins.
+ */
+public class ShareHandlingFragment extends Fragment {
+
+    /**
+     * Static factory method.
+     *
+     * Creates and attaches a new instance or returns a previously added one.
+     * This should be called early in whatever activity that declares the
+     * intent filters. Sibling fragments can then safely call this and rest
+     * assured that they are getting the same instance.
+     *
+     * @param fm **support** fragment manager.
+     * @return the fragment instance. The host activity should immediately
+     * follow this with a call to {@link #connect(HostConnection)}.
+     */
+    @NonNull
+    public static ShareHandlingFragment of(FragmentManager fm) {
+        Fragment fragment = fm.findFragmentByTag(TAG);
+        if (fragment == null) {
+            fragment = new ShareHandlingFragment();
+            fm.beginTransaction().add(fragment, TAG).commit();
+        }
+        return (ShareHandlingFragment) fragment;
+    }
+
+    private static final String TAG = ShareHandlingFragment.class.getCanonicalName();
+    private static final String KEY_HANDLED = "share.handled";
+    private static final String YOUTUBE_PREFIX = "plugin://plugin.video.youtube/play/?video_id=";
+    private static final String TWITCH_PREFIX = "plugin://plugin.video.twitch/playLive/";
+    private static final String VIMEO_PREFIX = "plugin://plugin.video.vimeo/play/?video_id=";
+    private static final String YOUTUBE_SHORT_URL = "://youtu\\.be/([^\\?\\s/]+)";
+    private static final String YOUTUBE_LONG_URL = "://(?:www\\.)?youtube\\.com/watch\\S*[\\?&]v=([^&\\s]+)";
+    private static final String TWITCH_URL = "://www\\.twitch\\.tv/([^\\?\\s/]+)";
+    private static final String VIMEO_URL = "://(?:www\\.|player\\.)?vimeo\\.com[^\\?\\s]*?/(\\d+)";
+
+    interface ShareHandler {
+        /**
+         * @return null when no url is recognized from this piece of data
+         * passed through the intent.
+         */
+        @Nullable String urlFrom(@Nullable String source);
+    }
+
+    interface OnFinish<T> {
+        void got(T result);
+    }
+
+    interface Task<T> {
+        void start(@NonNull OnFinish<? super T> then);
+    }
+
+    static class Seq<T> implements Task<T> {
```

Comment:
This should probably be moved into its own Class under org.xbmc.kore.utils as I think it will be applicable to other parts of Kore as well. Also `Seq` is a bit cryptic, how about naming this `Sequencer`?